Added CCA method

k-kagurazaka opened the pull request:
Hi, @venthur 

I added the Canonical Correlation Analysis (CCA) method to the processing.py.
I know the CCA is already implemented in the scikit-learn, but I think it is redundant.
The reason of why I re-implemented it, I want a simple CCA method for the SSVEP based BCI.
I would be glad if you could merge this pull request :)

k-kagurazaka committed:
Added CCA method

venthur commented:
Hi @k-kagurazaka thank you very much for your patch! I'll gladly merge it, but there are some minor issues that should be fixed first. I'll comment them in the appropriate lines of the patch.

May I ask in which group you work and how you got to know Wyrm?

Cheers and thank you very much again!

Basti

venthur commented on test/test_calculate_cca.py:
```diff
@@ -0,0 +1,67 @@
+        with self.assertRaises(AssertionError):
+            calculate_cca(dat, self.dat_y)
+
+    def test_calculate_csp_swapaxes(self):
```
copy-paste error. Should be `test_calculate_cca_swapaxes` (csp vs cca)

venthur commented on wyrm/processing.py:
```diff
@@ -1838,6 +1838,74 @@ def calculate_classwise_average(dat, classaxis=0):
+
+    This method calculates the canonical correlation and the corresponding weights.
+
+    Examples
```
Could you please re-order the blocks? The first one after the description text should be "Parameters", followed by "Returns", "Raises", "Examples", "References".

venthur commented on wyrm/processing.py:
```diff
@@ -1838,6 +1838,74 @@ def calculate_classwise_average(dat, classaxis=0):
+def calculate_cca(dat_x, dat_y, timeaxis=-2):
+    """Calculate the Canonical Correlation Analysis (CCA) of specified data.
+
+    This method calculates the canonical correlation and the corresponding weights.
```
Could you expand the long description just a little bit more? A few lines will be sufficient.

venthur commented on wyrm/processing.py:
```diff
@@ -1838,6 +1838,74 @@ def calculate_classwise_average(dat, classaxis=0):
+    http://en.wikipedia.org/wiki/Canonical_correlation
+
+    """
+    assert (len(dat_x.data.shape) == len(dat_y.data.shape) == 2 and
```
- Could you add a note in the above "Raises" block for the assertion error if data.shape != 2
- Could you add a unit test for the assertion?

k-kagurazaka replied:
In Wyrm, the continuous Data object usually consists of a 2d array, doesn't it?
So, I think that the case of 'data.shape != 2' is included the first condition and it's unit test is
already implemented at TestCalculateCCA.test_raise_error_with_non_continuous_data.

svendaehne commented on test/test_calculate_cca.py:
```diff
@@ -0,0 +1,67 @@
+        calculate_cca(self.dat_x, self.dat_y)
+        self.assertEqual(self.dat_x, cpy_x)
+        self.assertEqual(self.dat_y, cpy_y)
+
```
Can you please also test/demonstrate the functionality of your CCA implementation?
For this you would have to generate datasets X and Y which have at least one "component" in common. 

Pseudo-code for this could look like this:

``` python
N_COMPONENTS = 2

# create random input data in component space (a.k.a. source space)
Sx = randn(CHANNELS_X, SAMPLES)
Ax = randn(CHANNELS_X, CHANNELS_X)

Sy = randn(CHANNELS_Y, SAMPLES)
Ay = randn(CHANNELS_Y, CHANNELS_Y)

# make the first N components have the same time-course
Sy[0:N_COMPONENTS, :] = Sx[0:N_COMPONENTS, :]

# mix the components to generate the sensor space data
X = dot(Ax, Sx)
Y = dot(Ay, Sy)
# This concludes the generation of data
```

For testing, use CCA to compute the spatial filters and the canonical variables. The obtained canonical variables should be almost equal to the generated component time-courses, i.e. to first N_COMPONENTS columns of Sx and Sy, modulo sign and scaling. 

See also the unittest for the calculate_spoc function.

k-kagurazaka replied:
I think CCA cannot recovery source components from random mixed signals.

For testing the functionality of my CCA implementation, I considered a following case.
- X is a data set which includes a signal called S.
- Y is a data set which includes orthonormal basis whose linear span includes the S.

In this case, the first canonical correlation coefficient almost equals 1 and the canonical variable of Y almost equals S after zero-meaning and scaling.

Pseudo-code is here.

``` python
X = randn(SAMPLES, CHANNELS_X)
Y = randn(SAMPLES, CHANNELS_Y)
X[:, 0] = randn() * sin(2 * pi * FREQUENCY * t + randn() * pi) + randn()
Y[:, 0] = sin(2 * pi * FREQUENCY * t)
Y[:, 1] = cos(2 * pi * FREQUENCY * t)

rho, w_x, w_y = calculate_cca(X, Y)
# rho almost equals 1.0
# dot(Y, w_y) almost equals X[:, 0] after zero-meaning and scaling.
```

What do you think about this test?

svendaehne replied:
Hey Keita,

in prinicple CCA doesn't care about the temporal structure of the data.
I.e. it will work with random data just as fine as it will with
oscillations, as long as there really is a hidden correlation in the data.

I have prepared a little Matlab script that you can use as the basis for a
test. The first example in the script is similar to the case you described,
where you have one multivariate target signal (EEG data for example) as X
and then for Y the sine and cosine of the target frequency. The second case
in the script is the more general one, where the target signal is
completely random but it is mixed into both datasets X and Y. In both
cases, CCA should be able to extract the target signal from X and Y.

Here comes the Matlab code. If you don't have Matlab, it should be straight
forward to transform it into python code.
Cheers,
Sven

PS: the obtained correlations depend on the noise_level


%% parameters

CHANNELS_X = 15;
CHANNELS_Y = 17;
SAMPLES = 1000;
FREQUENCY = 10;

t = linspace(0,1,SAMPLES)';
noise_level = 0.1;

%% example 1 --> steady state BCI

% source signals for X
Sx = randn(SAMPLES, CHANNELS_X);
Sx(:,1) = sign(randn(1,1)) * sin(2 * pi * FREQUENCY * t + randn() * pi) +
noise_level*randn(size(t));

% mix the sources to obtain the measurement data
Ax = randn(CHANNELS_X,CHANNELS_X);
X = Sx * Ax;

% here only sin and cos of same frequency
Sy = [sin(2 * pi * FREQUENCY * t), cos(2 * pi * FREQUENCY * t)];
Y = Sy;

% perform CCA
[Wx,Wy,r] = canoncorr(X,Y);

sx = X * Wx(:,1);
sy = Y * Wy(:,1);

% correct for sign and scaling, then sx and sy should be almost equal
sx = zscore(sx);
sy = zscore(sy) * sign(sx'*sy);

figure
plot([sx, sy])

fprintf('r = %g\n',r(1))


%% example 2 --> general case

% source signals for X
Sx = randn(SAMPLES, CHANNELS_X);

% mix the sources to obtain the measurement data
Ax = randn(CHANNELS_X,CHANNELS_X);
X = Sx * Ax;

% source signals for Y
Sy = randn(SAMPLES, CHANNELS_Y);
% make the first source signal the same as in the X sources, perhaps plus
% some small amount of noise
Sy(:,1) = Sx(:,1) + noise_level*randn(size(t));

% mix the sources to obtain the measurement data
Ay = randn(CHANNELS_Y,CHANNELS_Y);
Y = Sy * Ay;

% perform
[Wx,Wy,r] = canoncorr(X,Y);

sx = X * Wx(:,1);
sy = Y * Wy(:,1);

% correct for sign and scaling, then sx and sy should be almost equal
sx = zscore(sx);
sy = zscore(sy) * sign(sx'*sy);

figure
plot([sx, sy])

fprintf('r = %g\n',r(1))







2015-01-07 14:46 GMT+01:00 Keita Kagurazaka <notifications@github.com>:

> In test/test_calculate_cca.py
> <https://github.com/venthur/wyrm/pull/1#discussion-diff-22586094>:
>
> > +    def test_calculate_csp_swapaxes(self):
> > +        """caluclate_cca must work with nonstandard timeaxis."""
> > +        res1 = calculate_cca(swapaxes(self.dat_x, 0, 1), swapaxes(self.dat_y, 0, 1), timeaxis=1)
> > +        res2 = calculate_cca(self.dat_x, self.dat_y)
> > +        np.testing.assert_array_equal(res1[0], res2[0])
> > +        np.testing.assert_array_equal(res1[1], res2[1])
> > +        np.testing.assert_array_equal(res1[2], res2[2])
> > +
> > +    def test_calculate_cca_copy(self):
> > +        """caluclate_cca must not modify argument."""
> > +        cpy_x = self.dat_x.copy()
> > +        cpy_y = self.dat_y.copy()
> > +        calculate_cca(self.dat_x, self.dat_y)
> > +        self.assertEqual(self.dat_x, cpy_x)
> > +        self.assertEqual(self.dat_y, cpy_y)
> > +
>
> I think CCA cannot recovery source components from random mixed signals.
>
> For testing the functionality of my CCA implementation, I considered a
> following case.
>
>    - X is a data set which includes a signal called S.
>    - Y is a data set which includes orthonormal basis whose linear span
>    includes the S.
>
> In this case, the first canonical correlation coefficient almost equals 1
> and the canonical variable of Y almost equals S after zero-meaning and
> scaling.
>
> Pseudo-code is here.
>
> X = randn(SAMPLES, CHANNELS_X)
> Y = randn(SAMPLES, CHANNELS_Y)
> X[:, 0] = randn() * sin(2 * pi * FREQUENCY * t + randn() * pi) + randn()
> Y[:, 0] = sin(2 * pi * FREQUENCY * t)
> Y[:, 1] = cos(2 * pi * FREQUENCY * t)
>
> rho, w_x, w_y = calculate_cca(X, Y)# rho almost equals 1.0# dot(Y, w_y) almost equals X[:, 0] after zero-meaning and scaling.
>
> What do you think about this test?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/venthur/wyrm/pull/1/files#r22586094>.
>

svendaehne replied:
Hey Keita,

in prinicple CCA doesn't care about the temporal structure of the data. I.e. it will work with random data just as fine as it will with oscillations, as long as there really is a hidden correlation in the data.

I have prepared a little Matlab script that you can use as the basis for a test. The first example in the script is similar to the case you described, where you have one multivariate target signal (EEG data for example) as X and then for Y the sine and cosine of the target frequency. The second case in the script is the more general one, where the target signal is completely random but it is mixed into both datasets X and Y. In both cases, CCA should be able to extract the target signal from X and Y.

Here comes the Matlab code. If you don't have Matlab, it should be straight forward to transform it into python code.
Cheers,
Sven

PS: the obtained correlations depend on the noise_level

``` matlab
%% parameters

CHANNELS_X = 15;
CHANNELS_Y = 17;
SAMPLES = 1000;
FREQUENCY = 10;

t = linspace(0,1,SAMPLES)';
noise_level = 0.1;

%% example 1 --> steady state BCI

% source signals for X
Sx = randn(SAMPLES, CHANNELS_X);
Sx(:,1) = sign(randn(1,1)) * sin(2 * pi * FREQUENCY * t + randn() * pi) + noise_level*randn(size(t));

% mix the sources to obtain the measurement data
Ax = randn(CHANNELS_X,CHANNELS_X);
X = Sx * Ax;

% here only sin and cos of same frequency
Sy = [sin(2 * pi * FREQUENCY * t), cos(2 * pi * FREQUENCY * t)];
Y = Sy;

% perform CCA
[Wx,Wy,r] = canoncorr(X,Y);

sx = X * Wx(:,1);
sy = Y * Wy(:,1);

% correct for sign and scaling, then sx and sy should be almost equal
sx = zscore(sx);
sy = zscore(sy) * sign(sx'*sy);

figure
plot([sx, sy])
fprintf('r = %g\n',r(1))

%% example 2 --> general case

% source signals for X
Sx = randn(SAMPLES, CHANNELS_X);

% mix the sources to obtain the measurement data
Ax = randn(CHANNELS_X,CHANNELS_X);
X = Sx * Ax;

% source signals for Y
Sy = randn(SAMPLES, CHANNELS_Y);
% make the first source signal the same as in the X sources, perhaps plus
% some small amount of noise
Sy(:,1) = Sx(:,1) + noise_level*randn(size(t));

% mix the sources to obtain the measurement data
Ay = randn(CHANNELS_Y,CHANNELS_Y);
Y = Sy * Ay;

% perform CCA
[Wx,Wy,r] = canoncorr(X,Y);

sx = X * Wx(:,1);
sy = Y * Wy(:,1);

% correct for sign and scaling, then sx and sy should be almost equal
sx = zscore(sx);
sy = zscore(sy) * sign(sx'*sy);

figure
plot([sx, sy])
fprintf('r = %g\n',r(1))
```

k-kagurazaka replied:
I'm sorry. I should have explained more thoroughly.
They are very specific cases that CCA can recover a source component from mixed signals, such as the mixture matrices are some unitary matrices.
In more wide cases such as the mixture matrices are regular, the canonical correlation coefficient has a tendency to show a high value.
However, the canonical variables aren't close to a source component in general.

k-kagurazaka replied:
Forgive me, @svendaehne 
I posted the above comment and pushed new commits before reading your new comment.
Would you wait that I will reflect your useful comment to my code.

svendaehne replied:
Sure, take your time. :-)

k-kagurazaka committed:
Fixed copy-paste error

k-kagurazaka committed:
Fixed document of processing.calculate_cca

k-kagurazaka committed:
Fixed processing.calculate_cca

Added zero-meaning of the specified signals before calculating
covariances. Added normalization of obtained weights.

k-kagurazaka committed:
Added the functionality test for processing.calculate_cca

k-kagurazaka commented:
Before fixing my test code, I'll answer your question.
I'm a grad student in Japan and belong a laboratory whose main theme is a brain signal processing.
In our lab, we are trying to migrate programming environments from MATLAB to python.
I searched python libraries for BCI and found your paper (http://arxiv.org/abs/1412.6378).

k-kagurazaka committed:
Changed to use eig instead of eigh

k-kagurazaka committed:
Changed the test of processing.calculate_cca

The new test uses random mixture matrices for checking the
functionality of the implementation of CCA.

k-kagurazaka commented:
I just pushed new commits which are reflected your advices.
I think there are too many commits in this pull-request, so do you want a squashed commit?
If yes, I'll closed this pull-request and request new one with the same content.

venthur commented:
Thanks for the new set of patches! No need to squash the commits, I'm happy to merge as it is. There is only one problem remaining: `reduce` is not available in Python3 anymore. You should replace it with `functools.reduce` which is available in Python2 and -3, an explicit for loop, or something similar.

Here is the log of the unittests which ran on Python3, check the end of the log:
https://s3.amazonaws.com/archive.travis-ci.org/jobs/46315764/log.txt

k-kagurazaka committed:
Replaced reduce to functools.reduce

k-kagurazaka committed:
Fixed a bug that type of weights is complex

Occasionally, obtained weights are complex vectors whose components have
the imaginary part with zero like "1.0 + j0".

k-kagurazaka commented:
Thanks @venthur .
I have no experience of Python 3.x, so I was at a loss.
I fixed the reduce problem and a bug described as a commit message.

venthur commented:
Thank you very much! Merged!

svendaehne commented:
Thanks from me as well. Good work!
